test/slim_check.lean

```diff
@@ -0,0 +1,75 @@
+
+import system.random.basic data.nat.prime
+
+instance fin.has_one' {n} [fact (0 < n)] : has_one (fin n) :=
+⟨ fin.of_nat' 1 ⟩
+
+instance fin.has_pow' {n} [fact (0 < n)] : has_pow (fin n) ℕ :=
+⟨ monoid.pow ⟩
+
+/-- fermat's primality test -/
+def primality_test (p : ℕ) (h : fact (0 < p)) : rand bool :=
+if h : 2 ≤ p-1 then do
+  n ← rand.random_r 2 (p-1), -- `random_r` requires a proof of `2 ≤ p-1` but it is dischared using `assumption`
+  return $ (fin.of_nat' n : fin p)^(p-1) = 1 -- we do arithmetic with `fin n` so that modulo and multiplication are interleaved
+else return (p = 2)
+
+def iterated_primality_test_aux (p : ℕ) (h : fact (0 < p)) : ℕ → rand bool
+| 0 := pure tt
+| (n+1) := do
+  b ← primality_test p h,
+  if b
+    then iterated_primality_test_aux n
+    else pure ff
+
+def iterated_primality_test (p : ℕ) : rand bool :=
+if h : 0 < p
+  then iterated_primality_test_aux p h 10
+  else pure ff
+
+def mfirst {α} {m} [monad m] (f : α → m bool) : list α → m (option α)
+| [] := pure none
+| (x :: xs) :=
+mcond (f x)
+  (pure x)
+  (mfirst xs)
```

Comment:
Move?

Reply:
There is a globally available mfirst but it depends on `m` implementing `alternative`. I could use it with the monad transformer `option_t` (which I might also have to define).

Reply:
Oh, I see. I guess it's fine to leave this here, but just note it's different from the globally available `mfirst`.

Reply:
Actually, `option_t` exists so I can just us it. The line that uses it becomes:

```lean
  some r ← option_t.run $ xs.mfirst (λ n, option_t.mk $ mcond (iterated_primality_test n) (pure (some n)) (pure none))
    | find_prime_aux n,
```